### PR TITLE
Implement forecast decomposition for `DeadlineMovingAverageModel`

### LINKED ISSUE #1171: Implement forecast decomposition for `DeadlineMovingAverageModel`
### 🚀 Feature Request

Implement forecast decomposition for `DeadlineMovingAverageModel`

### Proposal

1. In methods `forecast` and `predict` add flag `return_components: bool = False` (see base classes of models)
2. Implement target components extraction logic for case `return_components =True`:
- Components of forecast are target lags depending on `window` and `seasonality` parameters, there should be `window` components at all  
- Wights of the components are 1/`window`
- Names of the components are "target_component_lag_<lag_id>"
- Add target components to the result dataset via `add_target_component` method of the dataset
 

### Test cases

- Test components sum up to target for different parameter sets (`window`, `seasinality`)
- Test that the component names are correct for different parameter sets (`window`, `seasinality`)

### Additional context

Is blocked by Inference 2.1 milestone

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Closing issues

closes #1171

### CODE REVIEW

File: etna/models/deadline_ma.py [hunk 1]
```diff
@@ -200,7 +248,14 @@ def _forecast(self, df: pd.DataFrame, prediction_size: int) -> pd.DataFrame:
         df = df.iloc[-prediction_size:]
         y_pred = result_values[-prediction_size:]
         df.loc[:, pd.IndexSlice[:, "target"]] = y_pred
-        return df
+
+        components = None
```
Comment (alex-hse-repository) [inline etna/models/deadline_ma.py:252]: Let's rename it everywhere components, target_components  -> `target_components_df`

File: etna/models/deadline_ma.py [hunk 2]
```diff
@@ -200,7 +248,14 @@ def _forecast(self, df: pd.DataFrame, prediction_size: int) -> pd.DataFrame:
...
+        if return_components:
+            components = self._make_prediction_components(
+                result_template=result_template, context=result_template, prediction_size=prediction_size
```
Comment (alex-hse-repository) [inline etna/models/deadline_ma.py:255]: May you add comment here why do we use `result_template` as a context

File: etna/models/deadline_ma.py [hunk 3]
```diff
@@ -156,6 +164,47 @@ def _get_context_beginning(
...
+        for i in range(start_idx, end_idx):
+
+            obs_components = []
+            for w in range(1, self.window + 1):
+                prev_date = self._get_previous_date(date=result_template.index[i], offset=w)
+                obs_components.append(context.loc[prev_date].values)
+
+            components_data.append(obs_components)
+
+        raw_components = np.asarray(components_data, dtype=float)
```
Comment (alex-hse-repository) [inline etna/models/deadline_ma.py:194]: May you add comments with shapes here

File: etna/models/deadline_ma.py [hunk 4]
```diff
@@ -156,6 +164,47 @@ def _get_context_beginning(
...
+        raw_components = np.swapaxes(raw_components, -1, -2)
+        raw_components = raw_components.reshape(raw_components.shape[0], -1)
+        raw_components /= self.window
+
+        components = pd.DataFrame(data=raw_components)
```
Comment (alex-hse-repository) [inline etna/models/deadline_ma.py:199]: Just for the beauty of code
```
components = pd.DataFrame(data=raw_components, index=index[start_idx:end_idx])

segments_names = context.columns.get_level_values("segment") 
components_names = [f"target_component_{self.seasonality.name}_lag_{w}" for w in range(1, self.window + 1)]
components.columns = pd.MultiIndex.from_product(
    [segments_names, components_names], names=("segment", "feature")
)
```

File: tests/test_models/test_simple_models.py [hunk 5]
```diff
@@ -774,3 +794,48 @@ def test_sma_model_predict_components_correct(
...
+
+    method_to_call = getattr(model, method)
+    forecast = method_to_call(ts=long_periodic_ts, prediction_size=horizon, return_components=True)
+
+    assert sorted(forecast.target_components_names) == sorted(expected_components_names)
+
+
+@pytest.mark.parametrize("method", ("predict", "forecast"))
+@pytest.mark.parametrize(
+    "window,seasonality,expected_components_names",
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:822]: I guess you don't need this parametrization here

File: tests/test_models/test_simple_models.py [hunk 6]
```diff
@@ -55,6 +55,26 @@ def df():
     return tsds
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:55]: May you also add test checking the correctness of decomposition like in for SMA [here](https://github.com/tinkoff-ai/etna/blob/83db9b70cc5e7e9e7ad9e2b6cf2d1811a404900a/tests/test_models/test_simple_models.py#L787)

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_simple_models.py [hunk 7]
```diff
@@ -839,3 +839,19 @@ def test_deadline_ma_predict_components_sum_up_to_target(
     components = forecast.get_target_components()
 
     np.testing.assert_allclose(target.values, components.sum(axis=1, level="segment").values)
+
+
+@pytest.mark.parametrize(
+    "method_name, expected_values",
+    (("forecast", [[15, 2], [16, 4], [17, 6]]), ("predict", [[15, 2], [16, 4], [17, 6]])),
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:846]: Try to come up with the case where the expected values are different for forecast and predict -- i.e forecast should use predicted value as lag and predict uses real value

File: tests/test_models/test_simple_models.py [hunk 8]
```diff
@@ -839,3 +839,19 @@ def test_deadline_ma_predict_components_sum_up_to_target(
...
+)
+def test_deadline_ma_predict_components_correct(
+    simple_df, method_name, expected_values, window=1, seasonality="month", horizon=3
```
Comment (alex-hse-repository) [inline tests/test_models/test_simple_models.py:849]: May be also window=2

File: tests/test_models/test_simple_models.py (hunk 5 above)
Comment (brsnw250) [reply tests/test_models/test_simple_models.py:822]: All of this parameters are used.

Comment (brsnw250) [reply tests/test_models/test_simple_models.py]: Fixed

**Review by alex-hse-repository — APPROVED** [review]

**Pull request merged**